File: clients/libjabi/peripherals/can.cpp
#include <libjabi/byteorder.h>
#include <libjabi/interfaces/interface.h>

namespace jabi {

#include <jabi/peripherals.h>
#include <jabi/peripherals/can.h>

CANMessage::CANMessage()
:
    id(0), ext(false), fd(false), brs(false), rtr(false), len(0), data({})
{}

CANMessage::CANMessage(int id, int req_len, bool fd, bool brs)
:
    id(id), ext(id & ~0x7FF), fd(fd), brs(brs), rtr(true), len(req_len), data({})
{}

CANMessage::CANMessage(int id, std::vector<char> data, bool fd, bool brs)
:
    id(id), ext(id & ~0x7FF), fd(fd), brs(brs), rtr(false), len(data.size())
{
    if (data.size() > this->data.size()) {
        throw std::runtime_error("input data too long");
    }
    std::copy(data.begin(), data.end(), this->data.begin());
    for (auto i = data.size(); i < this->data.size(); i++) {
        this->data[i] = 0;
    }
}

std::ostream &operator<<(std::ostream &os, CANMessage const &m) {
    os << "CANMessage(id=" << m.id << ",ext=" << m.ext << ",fd=" << m.fd;
    os << ",brs=" << m.brs << ",rtr=" << m.rtr << ",len=" << m.len;
    if (!m.rtr) {
        os << ",data={";
        for (auto i = 0; i < m.len-1; i++) {
            os << static_cast<int>(std::byte(m.data[i])) << ",";
        }
        os << static_cast<int>(std::byte(m.data[m.len-1])) << "}";
    }
    os << ")";
    return os;
}

void Device::can_set_filter(int id, int id_mask, bool rtr, bool rtr_mask, int idx) {
    iface_req_t req = {
        .periph_id   = PERIPH_CAN_ID,
        .periph_idx  = static_cast<uint16_t>(idx),
        .periph_fn   = CAN_SET_FILTER_ID,
        .payload_len = sizeof(can_set_filter_req_t),
        .payload     = {0},
    };

    auto args = reinterpret_cast<can_set_filter_req_t*>(req.payload);
    args->id       = htole<uint32_t>(id);
    args->id_mask  = htole<uint32_t>(id_mask);
    args->rtr      = rtr;
    args->rtr_mask = rtr_mask;

    iface_resp_t resp = interface->send_request(req);
    if (resp.payload_len != 0) {
        throw std::runtime_error("unexpected payload length");
    }
}

void Device::can_set_rate(int bitrate, int bitrate_data, int idx) {
    iface_req_t req = {
        .periph_id   = PERIPH_CAN_ID,
        .periph_idx  = static_cast<uint16_t>(idx),
        .periph_fn   = CAN_SET_RATE_ID,
        .payload_len = sizeof(can_set_rate_req_t),
        .payload     = {0},
    };

    auto args = reinterpret_cast<can_set_rate_req_t*>(req.payload);
    args->bitrate      = htole<uint32_t>(bitrate);
    args->bitrate_data = htole<uint32_t>(bitrate_data);

    iface_resp_t resp = interface->send_request(req);
    if (resp.payload_len != 0) {
        throw std::runtime_error("unexpected payload length");
    }
}

void Device::can_set_mode(CANMode mode, int idx) {
    iface_req_t req = {
        .periph_id   = PERIPH_CAN_ID,
        .periph_idx  = static_cast<uint16_t>(idx),
        .periph_fn   = CAN_SET_STYLE_ID,
        .payload_len = sizeof(can_set_style_req_t),
        .payload     = {0},
    };

    auto args = reinterpret_cast<can_set_style_req_t*>(req.payload);
    args->mode = static_cast<uint8_t>(mode);

    iface_resp_t resp = interface->send_request(req);
    if (resp.payload_len != 0) {
        throw std::runtime_error("unexpected payload length");
    }
}

CANState Device::can_state(int idx) {
    iface_req_t req = {
        .periph_id   = PERIPH_CAN_ID,
        .periph_idx  = static_cast<uint16_t>(idx),
        .periph_fn   = CAN_STATE_ID,
        .payload_len = 0,
        .payload     = {0},
    };

    iface_resp_t resp = interface->send_request(req);
    if (resp.payload_len != sizeof(can_state_resp_t)) {
        throw std::runtime_error("unexpected payload length");
    }
    auto ret = reinterpret_cast<can_state_resp_t*>(req.payload);

    CANState state = {
        .state  = ret->state,
        .tx_err = ret->tx_err_cnt,
        .rx_err = ret->rx_err_cnt,
    };
    return state;
}

void Device::can_write(CANMessage msg, int idx) {
    iface_req_t req = {
        .periph_id   = PERIPH_CAN_ID,
        .periph_idx  = static_cast<uint16_t>(idx),
        .periph_fn   = CAN_WRITE_ID,
        .payload_len = sizeof(can_write_req_t),
        .payload     = {0},
    };

    auto args = reinterpret_cast<can_write_req_t*>(req.payload);
    args->id       = htole<uint32_t>(msg.id);
    args->id_type  = msg.ext;
    args->fd       = msg.fd;
    args->brs      = msg.brs;
    args->rtr      = msg.rtr;
    args->data_len = std::min<size_t>(msg.data.size(), msg.len);
    if (!msg.rtr) {
        memcpy(args->data, msg.data.data(), args->data_len);
        req.payload_len += args->data_len;
    }

    iface_resp_t resp = interface->send_request(req);
    if (resp.payload_len != 0) {
        throw std::runtime_error("unexpected payload length");
    }
}

int Device::can_read(CANMessage &msg, int idx) {
    iface_req_t req = {
        .periph_id   = PERIPH_CAN_ID,
        .periph_idx  = static_cast<uint16_t>(idx),
        .periph_fn   = CAN_READ_ID,
        .payload_len = 0,
        .payload     = {0},
    };

    iface_resp_t resp = interface->send_request(req);
    if (resp.payload_len == 0) {
        return -1; // empty buffer, no message returned
    }
    if (resp.payload_len < sizeof(can_read_resp_t)) {
        throw std::runtime_error("unexpected payload length");
    }

    auto ret = reinterpret_cast<can_read_resp_t*>(resp.payload);
    ret->num_left = letoh<uint16_t>(ret->num_left);
    ret->id       = letoh<uint32_t>(ret->id);

    if  ((ret->rtr && resp.payload_len != sizeof(can_read_resp_t)) ||
        (!ret->rtr && resp.payload_len != sizeof(can_read_resp_t) + ret->data_len) ||
          ret->data_len > msg.data.size()) {
        throw std::runtime_error("unexpected payload length");
    }

    msg.id  = ret->id;
    msg.ext = ret->id_type;
    msg.fd  = ret->fd;
    msg.brs = ret->brs;
    msg.rtr = ret->rtr;
    msg.len = ret->data_len;
    if (!ret->rtr) {
        memcpy(msg.data.data(), ret->data, ret->data_len);
    }
    return ret->num_left;
}

};

File: clients/jabi-rs/src/peripherals/can.rs
use crate::interfaces::InterfaceRequest;
use crate::Error;
use deku::{DekuContainerRead, DekuContainerWrite, DekuRead, DekuUpdate, DekuWrite};

pub enum CANMode {
    Normal = 0,
    Loopback = 1,
    ListenOnly = 2,
}

#[derive(DekuRead)] // happens to be the exact same format
#[deku(endian = "little")]
pub struct CANState {
    pub state: u8,
    pub tx_err: u8,
    pub rx_err: u8,
}

pub struct CANMessage {
    pub id: u32,
    pub ext: bool,
    pub fd: bool,
    pub brs: bool,
    pub rtr: bool,
    pub data: Vec<u8>,
}

impl std::fmt::Display for CANState {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "CANState(state={}, tx_err={}, rx_err={})",
            self.state, self.tx_err, self.rx_err
        )
    }
}

impl CANMessage {
    pub fn new(id: u32, data: Vec<u8>, fd_brs: bool) -> Self {
        Self {
            id: id,
            ext: id > 0x7FF,
            fd: fd_brs,
            brs: fd_brs,
            rtr: false,
            data: data,
        }
    }

    pub fn new_rtr(id: u32, len: usize, fd_brs: bool) -> Self {
        Self {
            id: id,
            ext: id > 0x7FF,
            fd: fd_brs,
            brs: fd_brs,
            rtr: true,
            data: vec![0; len],
        }
    }
}

impl std::fmt::Display for CANMessage {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "CANMessage(id={:#x}, ext={}, fd={}, brs={}, rtr={}, ",
            self.id, self.ext as u8, self.fd as u8, self.brs as u8, self.rtr as u8
        )?;
        if self.rtr {
            write!(f, "data.len()={}", self.data.len())?;
        } else {
            write!(f, "data={:?}", self.data)?;
        }
        write!(f, ")")
    }
}

const CAN_DATA_MAX_LEN: usize = 64;

enum Func {
    SetFilter = 0,
    SetRate = 1,
    SetMode = 2,
    State = 3,
    Write = 4,
    Read = 5,
}

fn gen_req(f: Func, idx: usize, payload: Vec<u8>) -> Result<InterfaceRequest, Error> {
    let mut r = InterfaceRequest {
        periph_id: crate::InstID::CAN as u16,
        periph_idx: idx as u16,
        periph_fn: f as u16,
        payload_len: 0,
        payload: payload,
    };
    r.update().map_or(Err(Error::PacketFormat), |_| Ok(r))
}

impl crate::Device {
    pub fn can_set_filter(
        &self,
        idx: usize, // Rust doesn't allow default args (yet)...
        id: u32,
        id_mask: u32,
    ) -> Result<(), Error> {
        #[derive(DekuWrite)]
        #[deku(endian = "little")]
        struct SetFilterRequest {
            id: u32,
            id_mask: u32,
        }
        let req = SetFilterRequest { id, id_mask };
        let resp = self.send(&gen_req(Func::SetFilter, idx, req.to_bytes().unwrap())?)?;
        if resp.len() == 0 {
            Ok(())
        } else {
            Err(Error::PacketFormat)
        }
    }

    pub fn can_set_rate(&self, idx: usize, bitrate: u32, bitrate_data: u32) -> Result<(), Error> {
        #[derive(DekuWrite)]
        #[deku(endian = "little")]
        struct SetRateRequest {
            bitrate: u32,
            bitrate_data: u32,
        }
        let req = SetRateRequest {
            bitrate,
            bitrate_data,
        };
        let resp = self.send(&gen_req(Func::SetRate, idx, req.to_bytes().unwrap())?)?;
        if resp.len() == 0 {
            Ok(())
        } else {
            Err(Error::PacketFormat)
        }
    }

    pub fn can_set_mode(&self, idx: usize, mode: CANMode) -> Result<(), Error> {
        #[derive(DekuWrite)]
        #[deku(endian = "little")]
        struct SetModeRequest {
            mode: u8,
        }
        let req = SetModeRequest { mode: mode as u8 };
        let resp = self.send(&gen_req(Func::SetMode, idx, req.to_bytes().unwrap())?)?;
        if resp.len() == 0 {
            Ok(())
        } else {
            Err(Error::PacketFormat)
        }
    }

    pub fn can_state(&self, idx: usize) -> Result<CANState, Error> {
        let resp = self.send(&gen_req(Func::State, idx, vec![])?)?;
        match CANState::from_bytes((&resp, 0)) {
            Ok(((s, _), ret)) if s.len() == 0 => Ok(ret),
            _ => Err(Error::PacketFormat),
        }
    }

    pub fn can_write(&self, idx: usize, msg: &CANMessage) -> Result<(), Error> {
        #[derive(DekuWrite)]
        #[deku(endian = "little")]
        struct WriteRequest<'a> {
            id: u32,
            id_type: u8,
            fd: u8,
            brs: u8,
            rtr: u8,
            data_len: u8,
            data: &'a [u8],
        }
        if msg.data.len() > CAN_DATA_MAX_LEN {
            return Err(Error::InvalidArgs);
        }
        let req = WriteRequest {
            id: msg.id,
            id_type: msg.ext as u8,
            fd: msg.fd as u8,
            brs: msg.brs as u8,
            rtr: msg.rtr as u8,
            data_len: msg.data.len() as u8,
            data: if msg.rtr { &[] } else { &msg.data },
        };
        let resp = self.send(&gen_req(Func::Write, idx, req.to_bytes().unwrap())?)?;
        if resp.len() == 0 {
            Ok(())
        } else {
            Err(Error::PacketFormat)
        }
    }

    pub fn can_read(&self, idx: usize) -> Result<Option<CANMessage>, Error> {
        #[derive(DekuRead)]
        #[deku(endian = "little")]
        struct ReadResponse {
            _num_left: u16, // "simple" interface, remains unused
            id: u32,
            id_type: u8,
            fd: u8,
            brs: u8,
            rtr: u8,
            data_len: u8,
            // data[]
        }
        let resp = self.send(&gen_req(Func::Read, idx, vec![])?)?;
        if resp.len() == 0 {
            Ok(None)
        } else {
            match ReadResponse::from_bytes((&resp, 0)) {
                Ok(((s, _), ret))
                    if ret.rtr != 0 || (ret.rtr == 0 && s.len() == ret.data_len.into()) =>
                {
                    Ok(Some(CANMessage {
                        id: ret.id,
                        ext: ret.id_type != 0,
                        fd: ret.fd != 0,
                        brs: ret.brs != 0,
                        rtr: ret.rtr != 0,
                        data: if ret.rtr != 0 {
                            vec![0; ret.data_len.into()]
                        } else {
                            s.to_vec()
                        },
                    }))
                }
                _ => Err(Error::PacketFormat),
            }
        }
    }
}
